Load admin monitors popup content via ajax

fqueiruga opened the pull request:
<!-- Comment: 
A great PR typically begins with the line below.
Replace XXXXX with the numeric part of the issue's id you created on JIRA.
Please note that if you want your changes backported into LTS, you will need to create a JIRA ticket for it. Read https://jenkins.io/download/lts/#backporting-process for more.
-->
See [JENKINS-XXXXX](https://issues.jenkins-ci.org/browse/JENKINS-XXXXX).

This PR aims to make loading the list of administrative monitors asynchronous and to not affect page load. This also has the benefit of fixing keyboard navigation through the page header.

No notable changes to show in screenshots, except the empty state 

<details>
<Summary>Screenshot of empty state</summary>

This should almost never be shown (for example, user gets permissions to read monitors removed and tries to load them before reloading the page).

<img width="410" alt="Captura de pantalla 2020-11-19 a las 17 09 21" src="https://user-images.githubusercontent.com/5738588/99692546-b6766d00-2a8a-11eb-92fd-6c5b94634e0f.png">

</details>

<!-- Comment: 
If the issue is not fully described in the ticket, add more information here (justification, pull request links, etc.).

 * We do not require JIRA issues for minor improvements.
 * Bugfixes should have a JIRA issue (backporting process).
 * Major new features should have a JIRA issue reference.
-->

### Proposed changelog entries

* Load the administrative monitors popup on demand to reduce page load time.
* Fixes a bug where the page header would not be navigable using keyboard when there were active administrative monitors.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@timja
@daniel-beck 
@Wadeck 
@oleg-nenashev

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

fqueiruga committed:
Admin monitors popup is now loaded through an API

fqueiruga committed:
Fixes form submission inside the popup

- Wraps it in an l:ajax template so that proper variables are set
- Applies the csrf crumbs to the forms after the popup is constructed

fqueiruga committed:
Handle authorization and empty state

timja commented:
> Fixes a bug where the page header would not be navigable using keyboard when there were active administrative monitors.

nice! (if you keep tabbing you can navigate but there's some invisible elements to get through first)

fqueiruga commented:
Exactly that one

timja commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/footer.jelly:
```diff
@@ -23,73 +23,65 @@ THE SOFTWARE.
             (function() {
-                var amContainer = document.getElementById("visible-am-container");
-                var amInsertion = document.getElementById("visible-am-insertion");
+            var amContainer = document.getElementById("visible-am-container");
```
looks like the indent change should be reverted here?

fqueiruga replied:
YES

timja commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/resources.js:
```diff
@@ -1,78 +1,90 @@
-        if (el === list) {
-            return; // clicked in the list
+        function onEscClose(e) {
+            if (e.keyCode === 27) {
```
would be good to extract this to a variable so that people who don't know keycodes know what this is

fqueiruga replied:
I put a comment next to it to clarify, if that's OK.

timja reviewed (commented):
looks really good, only minor comments

timja commented:
> Exactly that one, shouldn't happen anymore

yes, tested that, and it works now

timja commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsApiData/monitorsList.properties:
```diff
@@ -0,0 +1 @@
+no_active_monitors=There are currently no active monitors
```
what about this?
```suggestion
no_active_monitors=There are currently no active warnings
```

fqueiruga committed:
Fix JS error where there were monitors of only one kind

fqueiruga committed:
Revert changes to indentation of inline JS

fqueiruga committed:
Apply suggestions from code review

Co-authored-by: Tim Jacomb <timjacomb1+github@gmail.com>

fqueiruga committed:
Clarify ESC keyCode with a comment

timja commented on core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/resources.js:
```diff
@@ -1,78 +1,94 @@
-        if (el === list) {
-            return; // clicked in the list
+        function onEscClose(e) {
+            if (e.keyCode === 27) { // If the key press is ESCAPE...
```
I meant something like this, what do you think?
```suggestion
            var escapeKeyCode = 27;
            if (e.keyCode === escapeKeyCode) {
```

fqueiruga committed:
Update core/src/main/resources/jenkins/management/AdministrativeMonitorsDecorator/resources.js

Co-authored-by: Tim Jacomb <timjacomb1+github@gmail.com>

fqueiruga commented:
Seems that the ATH tests are being aborted. Any idea what may be happening?

fcojfernandez reviewed (approved):
At some point the administrative monitors should be moved to the header from the footer IMO, but as a first step it's great

oleg-nenashev commented on core/src/main/java/jenkins/management/AdministrativeMonitorsApi.java:
```diff
@@ -0,0 +1,62 @@
+@Extension
+@Restricted(NoExternalUse.class)
+public class AdministrativeMonitorsApi implements RootAction {
+    @GET
```
Should these calls have explicit permission checks?

fqueiruga replied:
The authorization is implicit through the decorator. That said I think it would make sense to have more explicit auth. I have no idea how to got about this. If anyone wants to commit straight to my branch please go ahead :D

fqueiruga replied:
@daniel-beck @Wadeck  how would you go about this?

timja replied:
it's quite common to rely on permission checks in this way as far as I know

Wadeck replied:
In this case it's a bit special because the permission is checked deeper in the code. It was recently changed from Admin to SystemRead. It means that if we have an additional check here, it's a bit a duplicate code, so I am fine with the current approach.

Of course if the method is starting to do something else, the explicit permission will be required, but we should not be future-proof without good rational.

fqueiruga commented:
Any chance this could make it this week? :D

oleg-nenashev reviewed (approved):
Should be good to go.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

Wadeck commented:
Please do not merge, I am working on some quick changes here to prevent "regression" from the admin monitor separation in terms of performance.

Wadeck committed:
Multiple small adjustments

- Remove unused imports
- Remove non-necessary call for the response that are set in the getView
- Keep the previous improvement about limiting the number of call to getMonitorList()

Wadeck commented:
👍 really nice job @fqueiruga :)

I adjusted a bit the code and added non-regression tests. I did not expect someone to remove that "feature" for the performance aspect, so now it's covered ;) In your PR the page was asking 3x the getMonitorList.

After some PRs like this one you will no longer be able to say you're not a Java developer ;)

Wadeck reviewed (approved):
Test added + rollback the filtering removal to keep the 1x call instead of 3x

fqueiruga commented on core/src/main/java/jenkins/management/AdministrativeMonitorsApi.java:
```diff
@@ -12,28 +11,19 @@
     @GET
     public void doNonSecurityPopupContent(StaplerRequest req, StaplerResponse resp) throws IOException, ServletException {
-        resp.setStatus(HttpServletResponse.SC_OK);
-        resp.setContentType("text/html");
```
why aren't this needed?

Wadeck replied:
It's done automatically by the JellyFacet (need to find exactly the code, inside Stapler, digging...)

oleg-nenashev reviewed (approved):
Looks good. Just in case I checked improper usages of the restricted APIs, all looks good. 🚢 🇮🇹

daniel-beck commented:
This PR causes:

```
WARNING	j.s.s.StaplerFilteredActionListener#onFieldTrigger: New Stapler routing rules result in the URL "/administrativeMonitors/securityPopupContent" no longer being allowed. If you consider it safe to use, add the following to the whitelist: "field jenkins.model.Jenkins administrativeMonitors". Learn more: https://jenkins.io/redirect/stapler-routing
```

Due to the naming collision between the field in Jenkins https://github.com/jenkinsci/jenkins/blob/9a8588951b77a938fcf4ba9cee68d4c3e98d28b0/core/src/main/java/jenkins/model/Jenkins.java#L722 and the `getDynamic` fallback to `RootAction` https://github.com/jenkinsci/jenkins/blob/9a8588951b77a938fcf4ba9cee68d4c3e98d28b0/core/src/main/java/jenkins/model/Jenkins.java#L3787

daniel-beck commented:
Probably caused [JENKINS-70036](https://issues.jenkins.io/browse/JENKINS-70036).
